hbase-testing-util/src/main/java/org/apache/hadoop/hbase/testing/ServerTestingHBaseCluster.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.testing;
+
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.HBaseTestingUtil;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * This interface is intended for uncommon testing use cases which need to access internal state
+ * on an HBase minicluster. This may be necessary when testing code which runs within an HBase
+ * server process, such as coprocessors or replication endpoints. It provides access to the
+ * underlying {@link HBaseTestingUtil}, which its parent {@link TestingHBaseCluster} interface
+ * intentionally does not. External use cases that just need to test their code using the HBase
+ * client API should use {@link TestingHBaseCluster} instead.
+ */
+@InterfaceAudience.LimitedPrivate(
+  {HBaseInterfaceAudience.COPROC, HBaseInterfaceAudience.REPLICATION,
+    HBaseInterfaceAudience.PHOENIX})
+public interface ServerTestingHBaseCluster extends TestingHBaseCluster {
+  /**
+   *
+   * @return the utility running the minicluster
+   */
+  HBaseTestingUtil getUtil();
```

Comment:
This is a no no.

Let's find another way to only expose the necessary methods for testing coprocessors and other server side hooks such as tags. The point here is, HBaseTestingUtil is a class, not an interface, which makes very hard for the HBase developper to make it stable while keeping it clean.

Reply:
@Apache9 - thanks, that's a good point. In the next revision of this patch, I will create an interface for HBTU (and probably its superclasses) so this interface doesn't return a concrete class. I'm going to be on vacation for about a week but will take this up when I get back.